### PR TITLE
LXD auto detect

### PR DESCRIPTION
**Pull request by SimonRichardson**
## Description of change

The following is a work in progress and has only been manually tested.
Essentially the idea goes, that if you already have an existing setup
and just want juju to take over, with the previsor that you already
have the certs and config setup correctly, then this just works.

The down side to this setup is two fold:

 1. We touch a LOT of surface area to get the certificates to be
 detected at every point (bootstrap, destroy, kill)
 1. The credentials are not stored correctly, as we have to touch
 destroy and kill commands, which feels wrong.

This is a demo, I'm unsure if we should continue, but it's interesting
that it works nicely using the lxc config.

Note: there are no tests, as I would like to get feedback before I
spend any more time on this.

## QA steps

This assumes that you've connected to a remote lxd node using a trust
password with your local lxc and it's configured the creds correctly.

```sh
juju bootstrap nuc8 --credential=nuc8
```

You can also destroy the controller.

```sh
juju destroy-controller nuc8-default --destroy-all-models
```

## Documentation changes

The following allows onboarding of existing lxc users, who have a
local setup, easier migration to juju.

## Bug reference

N/A

### CODE REVIEW

**Comment by manadart** [comment]
It doesn't look like you can rely on ~/.config/lxc/config.yml.

With Snap-installed LXD, the location must be different because I have remotes that are not in that file.

**Comment by SimonRichardson** [comment]
@manadart good shout, I'll have a look at snap locations.

**Comment by SimonRichardson** [comment]
@manadart so as long as we use the `$HOME` env var as the root for the path, then it's fine. As the `$HOME` env var in snap is `/home/ubuntu/snap/lxd/$revsion` (surprised that's not current tbh). So the code is correct.

https://github.com/juju/juju/pull/8946/files#diff-e8543713fc0c30ea33131d41edf815e7R183

**Comment by SimonRichardson** [comment]
https://trello.com/c/bzEF5Lu8/368-auto-detect-existing-lxc-server-configurations-improve-ux-for-onboarding

**Comment by SimonRichardson** [comment]
I've got feedback offline on this, so we need to store the information inside `juju` bounds and then we won't need the `destroy` and `kill` changes.

File: provider/lxd/credentials.go [hunk 1]
```diff
@@ -135,30 +136,76 @@ func (environProviderCredentials) CredentialSchemas() map[cloud.AuthType]cloud.C
...
+
 	const credName = lxdnames.DefaultCloud
 	label := fmt.Sprintf("LXD credential %q", credName)
 	certCredential, err := p.finalizeLocalCredential(
 		ioutil.Discard, svr, string(certPEM), string(keyPEM), label,
 	)
+	return certCredential, errors.Trace(err)
+}
+
+func (p environProviderCredentials) detectRemoteCredentials(certPEM, keyPEM []byte) (map[string]cloud.Credential, error) {
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply provider/lxd/credentials.go]: Agreed, I'll change.

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**